service/frontend/adminHandler.go

```diff
@@ -1479,6 +1483,173 @@ func (adh *AdminHandler) GetTaskQueueTasks(
 	}, nil
 }
 
+func (adh *AdminHandler) DeleteWorkflowExecution(
+	ctx context.Context,
+	request *adminservice.DeleteWorkflowExecutionRequest,
+) (_ *adminservice.DeleteWorkflowExecutionResponse, err error) {
+	defer log.CapturePanic(adh.logger, &err)
+	scope, sw := adh.startRequestProfile(metrics.AdminDeleteWorkflowExecutionScope)
+	defer sw.Stop()
+
+	if request == nil {
+		return nil, adh.error(errRequestNotSet, scope)
+	}
+
+	if err := validateExecution(request.Execution); err != nil {
+		return nil, adh.error(err, scope)
+	}
+
+	namespaceID, err := adh.namespaceRegistry.GetNamespaceID(namespace.Name(request.GetNamespace()))
+	if err != nil {
+		return nil, adh.error(err, scope)
+	}
+	execution := request.Execution
+
+	shardID := common.WorkflowIDToHistoryShard(
+		namespaceID.String(),
+		execution.GetWorkflowId(),
+		adh.numberOfHistoryShards,
+	)
+
+	if execution.RunId == "" {
+		resp, err := adh.persistenceExecutionManager.GetCurrentExecution(ctx, &persistence.GetCurrentExecutionRequest{
+			ShardID:     shardID,
+			NamespaceID: namespaceID.String(),
+			WorkflowID:  execution.WorkflowId,
+		})
+		if err != nil {
+			return nil, adh.error(err, scope)
+		}
+		execution.RunId = resp.RunID
+	}
+
+	var branchTokens [][]byte
+	var startTime, closeTime *time.Time
+	cassVisBackend := strings.Contains(adh.visibilityMgr.GetName(), cassandra.CassandraPersistenceName)
+
+	resp, err := adh.persistenceExecutionManager.GetWorkflowExecution(ctx, &persistence.GetWorkflowExecutionRequest{
+		ShardID:     shardID,
+		NamespaceID: namespaceID.String(),
+		WorkflowID:  execution.WorkflowId,
+		RunID:       execution.RunId,
+	})
+	if err != nil {
+		if common.IsContextCanceledErr(err) || common.IsContextDeadlineExceededErr(err) {
+			return nil, adh.error(err, scope)
+		}
+		// continue to deletion
+		adh.logger.Warn("Unable to load mutable state when deleting workflow execution, "+
+			"will skip deleting workflow history and cassandra visibility record", tag.Error(err))
+	} else {
+		// load necessary information from mutable state
+		executionInfo := resp.State.GetExecutionInfo()
+		histories := executionInfo.GetVersionHistories().GetHistories()
+		branchTokens = make([][]byte, 0, len(histories))
+		for _, historyItem := range histories {
+			branchTokens = append(branchTokens, historyItem.GetBranchToken())
+		}
+
+		if cassVisBackend {
+			if resp.State.ExecutionState.State != enumsspb.WORKFLOW_EXECUTION_STATE_COMPLETED {
+				startTime = executionInfo.GetStartTime()
+			} else {
+				completionEvent, err := adh.getWorkflowCompletionEvent(ctx, shardID, resp.State)
```

Comment:
We probably should store close time in executionInfo as well.

Reply:
yeah. @yux0 have a pr for that actually. I am not sure why we decided to close it in the end.
https://github.com/temporalio/temporal/pull/2709